fendermint/ipc/src/parent.rs

```diff
@@ -0,0 +1,202 @@
+        self.read_cache(|cache| cache.membership.clone())
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) {
+        let mut cache = self.cache.write().unwrap();
+        cache.block_hash.remove_key_till(finality.height);
+    }
+}
+
+impl Clone for PollingParentSyncer {
+    fn clone(&self) -> Self {
+        Self {
+            config: self.config.clone(),
+            started: Arc::new(AtomicBool::new(false)),
+            cache: self.cache.clone(),
+            ipc_agent_proxy: self.ipc_agent_proxy.clone(),
+            handle: None,
+        }
+    }
+}
+
+struct ParentSyncerCache {
+    block_hash: RangeKeyCache<BlockHeight, Vec<u8>>,
+    top_down_message: RangeKeyCache<Nonce, CrossMsg>,
+    membership: Option<ValidatorSet>,
+}
+
+type LockedCache = Arc<RwLock<ParentSyncerCache>>;
+
+impl PollingParentSyncer {
+    fn read_cache<F, T>(&self, f: F) -> T
+    where
+        F: Fn(&ParentSyncerCache) -> T,
+    {
+        let cache = self.cache.read().unwrap();
+        f(&cache)
+    }
+
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let parent_syncer = self.ipc_agent_proxy.clone();
+        let config = self.config.clone();
+        let cache = self.cache.clone();
+
+        let handle =
+            tokio::spawn(async move { sync_with_parent(config, parent_syncer, cache).await });
+        self.handle = Some(handle);
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent(
+    config: Config,
+    agent_proxy: Arc<AgentProxy>,
+    lock: LockedCache,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        let hashes = fetch_block_hashes(&config, &agent_proxy, &lock).await?;
+        update_top_down_msgs(&config, &agent_proxy, &lock).await?;
+        update_membership(&config, &agent_proxy, &lock).await?;
+
+        let mut cache = lock.write().unwrap();
+        for r in hashes {
+            cache.block_hash.insert_after_lower_bound(r.0, r.1);
+        }
+    }
+}
+
+async fn update_top_down_msgs(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<Vec<CrossMsg>> {
+    Ok(vec![])
+}
+
+async fn update_membership(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<ValidatorSet> {
+    Ok(ValidatorSet::default())
+}
+
+async fn fetch_block_hashes(
+    config: &Config,
+    agent_proxy: &Arc<AgentProxy>,
+    lock: &LockedCache,
+) -> anyhow::Result<Vec<(BlockHeight, Vec<u8>)>> {
+    let latest_height = match agent_proxy.get_chain_head_height().await {
```

Comment:
Would it be technically possible for `get_chain_head_height` to return a _lower_ height than what came before? 

Say the parent was a longest chain and it was undergoing a re-org, that is, it was switching from a shorter to a longer fork, and first it rolled back the shorter fork to the greatest common ancestor, then forward on the new new fork. If we queried during rollback, it might appear to go backwards, and then it would leave some invalid block hashes in the cache. 

For example if we had this:
```
B0 - B1 - B2 - B3 - B4
   \ 
     B1' - B2' - B3' - B4' - B5'
```
1. Our client is at B4, latest height is 4
2. The parent decides to roll back to B0 and then forward to B5'
3. We query it during the rollback when it's at B3', we fetch B1', B2', B3'
4. We insert these into the cache, which now contains B0, B1', B2', B3', B4

B4 is not a descendant of B3' but it was left there because it's not included in the catch.

If you included the clearing out of the cache above the the new minimum, rather than just do inserts, this problem went away.

Reply:
Good point raise, will take note of this one.